pkg/networkpath/traceroute/tcp/utils.go

```diff
@@ -280,108 +280,83 @@ func parseIPv4Layer(pkt gopacket.Packet) (net.IP, net.IP, error) {
+	if header.Protocol != 1 || header.Version != 4 ||
+		header.Src == nil || header.Dst == nil {
+		log.Errorf("invalid IP header for ICMP packet")
+		return nil, fmt.Errorf("invalid IP header for ICMP packet: %+v", header)
+	}
+	icmpResponse.SrcIP = header.Src
+	icmpResponse.DstIP = header.Dst
+
+	var icmpv4Layer layers.ICMPv4
+	decoded := []gopacket.LayerType{}
+	icmpParser := gopacket.NewDecodingLayerParser(layers.LayerTypeICMPv4, &icmpv4Layer)
+	icmpParser.IgnoreUnsupported = true // ignore unsupported layers, we will decode them in the next step
+	if err := icmpParser.DecodeLayers(payload, &decoded); err != nil {
 		return nil, fmt.Errorf("failed to decode ICMP packet: %w", err)
 	}
-
-	if icmpLayer := pkt.Layer(layers.LayerTypeICMPv4); icmpLayer != nil {
-		icmp, ok := icmpLayer.(*layers.ICMPv4)
-		if !ok {
-			return nil, fmt.Errorf("failed to assert ICMPv4 layer type")
-		}
-		icmpResponse.TypeCode = icmp.TypeCode
-
-		var payload []byte
-		if len(icmp.Payload) < 40 {
-			log.Tracef("Payload length %d is less than 40, extending...\n", len(icmp.Payload))
-			payload = make([]byte, 40)
-			copy(payload, icmp.Payload)
-			// we have to set this in order for the TCP
-			// parser to work
-			payload[32] = 5 << 4 // set data offset
-		} else {
-			payload = icmp.Payload
-		}
-
-		// if we're in an ICMP packet, we know that we should have
-		// an inner IPv4 and TCP header section
-		var innerIPLayer layers.IPv4
-		var innerTCPLayer layers.TCP
-		decoded := []gopacket.LayerType{}
-		innerIPParser := gopacket.NewDecodingLayerParser(layers.LayerTypeIPv4, &innerIPLayer, &innerTCPLayer)
-		if err := innerIPParser.DecodeLayers(payload, &decoded); err != nil {
-			return nil, fmt.Errorf("failed to decode ICMP payload: %w", err)
-		}
-		icmpResponse.InnerSrcIP = innerIPLayer.SrcIP
-		icmpResponse.InnerDstIP = innerIPLayer.DstIP
-		icmpResponse.InnerSrcPort = uint16(innerTCPLayer.SrcPort)
-		icmpResponse.InnerDstPort = uint16(innerTCPLayer.DstPort)
-		icmpResponse.InnerSeqNum = innerTCPLayer.Seq
-	} else {
-		return nil, fmt.Errorf("packet does not contain an ICMP layer")
+	// since we ignore unsupported layers, we need to check if we actually decoded
+	// anything
+	if len(decoded) < 1 {
+		return nil, fmt.Errorf("failed to decode ICMP packet, no layers decoded")
 	}
-
-	return &icmpResponse, nil
-}
-
-func parseTCP(header *ipv4.Header, payload []byte) (*tcpResponse, error) {
-	packetBytes, err := MarshalPacket(header, payload)
-	if err != nil {
-		return nil, fmt.Errorf("failed to marshal packet: %w", err)
+	icmpResponse.TypeCode = icmpv4Layer.TypeCode
+
+	var icmpPayload []byte
+	if len(icmpv4Layer.Payload) < 40 {
+		log.Tracef("Payload length %d is less than 40, extending...\n", len(icmpv4Layer.Payload))
+		icmpPayload = make([]byte, 40)
+		copy(icmpPayload, icmpv4Layer.Payload)
+		// we have to set this in order for the TCP
+		// parser to work
+		icmpPayload[32] = 5 << 4 // set data offset
+	} else {
+		icmpPayload = icmpv4Layer.Payload
```

Comment:
Isn't the layer payload a pointer into the packet buffer and so isn't guaranteed to not change?

Reply:
Ah, you're only using this below, so this should be fine.